vm_phys.c

```diff
@@ -0,0 +1,251 @@
+    page_buddy->order = order-1;
+    page_buddy->size = ORDER_SIZE(page_buddy->order)*PAGE_SIZE;
+
+    TAILQ_INSERT_HEAD(&seg->free_queues[order-1], page, freeq);
+    TAILQ_INSERT_HEAD(&seg->free_queues[order-1], page_buddy, freeq);
+}
+
+
+static void vm_phys_reserve_from_seg(struct vm_phys_seg *seg, vm_paddr_t start, vm_paddr_t end)
+{
+    int i = 0;
+    /* Split segment into smaller segments, 
+     * remove them from free queue, mark space as reserved */
+    for(i = VM_NFREEORDER-1; i >= 0; i--) {
+        struct vm_page *pg_ptr;
+        struct vm_page *pg_it = TAILQ_FIRST(&seg->free_queues[i]);
+        while(pg_it) {
+            if(pg_it->phys_addr >= start && 
+                pg_it->phys_addr + pg_it->size <= end) {
+                pg_ptr = pg_it;
+                while(pg_ptr++ < pg_it + ORDER_SIZE(pg_it->order))
+                    pg_ptr->flags &= VM_RESERVED;
+                TAILQ_REMOVE(&seg->free_queues[i], pg_it, freeq);
+                /* Go to start of list because it's corrupted */
+                pg_it = TAILQ_FIRST(&seg->free_queues[i]);
+                continue;
+            }
+            else 
+                if(  (pg_it->phys_addr+pg_it->size > start &&
+                      pg_it->phys_addr < start) ||
+                     (pg_it->phys_addr+pg_it->size > end &&
+                      pg_it->phys_addr < end)) {
+                split_page(seg, pg_it);
+                /* Go to start of list because it's corrupted */
+                pg_it = TAILQ_FIRST(&seg->free_queues[i]);
+                continue;
+            }
+            pg_it = TAILQ_NEXT(pg_it, freeq);
+        }
+    }
+}
+
+void vm_phys_reserve(vm_paddr_t start, vm_paddr_t end) {
+    
+    struct vm_phys_seg *seg_it;
+    TAILQ_FOREACH(seg_it, &seglist, segq) {
+        if(seg_it->start <= start && seg_it->end >= end) {
+            vm_phys_reserve_from_seg(seg_it, start, end);
+            return;
+        }
+    }
+
+}
+
+static struct vm_page * vm_phys_alloc_from_seg(struct vm_phys_seg *seg, size_t order) {
+    struct vm_page *res = NULL;
+    size_t ord = order; 
+    /* lowest non-empty order higer than order */
+    while(TAILQ_EMPTY(&seg->free_queues[ord]) && ord < VM_NFREEORDER)
+        ord++;
+
+    if(ord == VM_NFREEORDER)
+        return NULL;
+
+    while(TAILQ_EMPTY(&seg->free_queues[order])) {
+        split_page(seg, TAILQ_FIRST(&seg->free_queues[ord]));
+        ord--;
+    }
+
+    res = TAILQ_FIRST(&seg->free_queues[order]);
+    TAILQ_REMOVE(&seg->free_queues[order], res, freeq);
+
+    res->size = ORDER_SIZE(res->order)*PAGE_SIZE;
+    return res;
+}
+
+struct vm_page * vm_phys_alloc(size_t order) {
+
+    struct vm_phys_seg *seg_it;
+    TAILQ_FOREACH(seg_it, &seglist, segq) {
+        struct vm_page *res;
+        res = vm_phys_alloc_from_seg(seg_it, order);
+        if(res != NULL) return res;
+    }
+
+    panic("vm_phys_alloc: cannot allocate page of given order");
```

Comment:
Ekhm... if this function returns `NULL` then a swapper (that we don't have yet) should be launched.